libs/esp32boot/esp32init.erl

```diff
@@ -0,0 +1,394 @@
+% distributed under the License is distributed on an "AS IS" BASIS,
+% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+% See the License for the specific language governing permissions and
+% limitations under the License.
+%
+% SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+%
+
+-module(esp32init).
+
+-export([
+    start/0,
+    start_repl/1,
+    start_network/0,
+    maybe_start_network/0,
+    handle_req/3,
+    erase_net_config/0,
+    save_net_config/2
+]).
+
+-record(nc_state, {socket, pending_pid, pending_ref}).
+
+-define(DEFAULT_AP_SSID, <<"AtomVM-ESP32">>).
+-define(DEFAULT_AP_PSK, <<"esp32default">>).
+-define(DEFAULT_CONSOLE_PORT, 2323).
+-define(DEFAULT_WEB_SERVER_PORT, 8080).
+
+start() ->
+    erlang:display("AtomVM init."),
+
+    io:format("Starting application...~n"),
+
+    Exit =
+        try boot() of
+            Result -> {exit, Result}
+        catch
+            Error -> {crash, Error}
+        end,
+    erlang:display(Exit),
+
+    io:format("Looping...~n"),
+
+    loop().
+
+loop() ->
+    receive
+        Msg ->
+            erlang:display({received_message, Msg}),
+            loop()
+    end.
+
+maybe_start_dev_mode(SystemStatus) ->
+    case {SystemStatus, esp:nvs_get_binary(atomvm, dev_mode)} of
+        {_, <<"always">>} ->
+            start_dev_mode();
+        {_, <<"never">>} ->
+            not_started;
+        {ok, undefined} ->
+            not_started;
+        {failed_app_start, undefined} ->
+            start_dev_mode()
+    end.
+
+start_dev_mode() ->
+    avm_pubsub:start(default_pubsub),
+    spawn(?MODULE, maybe_start_network, []),
+    started.
+
+%%
+%% Boot handling
+%%
+
+% TODO: add support for multiple apps
+% /dev/partition/by-name/app1.avm
+% /dev/partition/by-name/app2.avm
+
+boot() ->
+    BootPath = get_boot_path(),
+    case atomvm:add_avm_pack_file(BootPath, [{name, app}]) of
+        ok ->
+            StartModule = get_start_module(),
+            maybe_start_dev_mode(ok),
+            StartModule:start();
+        {error, Reason} ->
+            io:format("Failed app start: ~p.~n", [Reason]),
+            maybe_start_dev_mode(failed_app_start)
+    end.
+
+get_boot_path() ->
+    case esp:nvs_get_binary(atomvm, boot_path) of
+        undefined ->
+            "/dev/partition/by-name/main.avm";
+        Path ->
+            Path
+    end.
+
+get_start_module() ->
+    case esp:nvs_get_binary(atomvm, start_module) of
+        undefined ->
+            case atomvm:get_start_beam(app) of
+                error ->
+                    main;
+                {ok, ModuleNameWithExt} when is_binary(ModuleNameWithExt) ->
+                    Len = byte_size(ModuleNameWithExt) - byte_size(<<".beam">>),
+                    ModuleName = binary:part(ModuleNameWithExt, 0, Len),
+                    erlang:binary_to_atom(ModuleName, latin1)
+            end;
+        Module ->
+            erlang:binary_to_atom(Module, latin1)
+    end.
+
+%%
```

Comment:
network, web, REPL logic 👇🏽  should all go in a separate modules.

Reply:
short story: `pack_runnable` accepts just one module and I wasn't planning to go deep in the cmake rabbit hole with this PR.
as a temporary workaround we might agree to:
- move ALISP and webserver to a separated archive: `esp32devmode`
- `pack_runnable(esp32boot esp32init esp32devmode [...])`

Does it make sense for you?

Reply:
I used a work around, I'll open an issue about cmake macros.